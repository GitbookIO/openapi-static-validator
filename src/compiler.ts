import escodegen from 'escodegen';
import { namedTypes, builders } from 'ast-types';
import { RequestErrorClass, ValidationErrorClass, ValidationErrorIdentifier } from './error';
import { OpenAPIRef, OpenAPISpec } from './types';
import { compileValueSchema } from './compileValueSchema';
import { hash } from './hash';
import { compileValidateRequest } from './compileValidateRequest';

/**
 * Compiler for OpenAPI specs.
 */
export class Compiler {
    private input: OpenAPISpec;

    /** Map of hash to the identifier */
    private hashes: Map<string, string> = new Map();
    /** Map of objects to the identifier */
    private objectHashes: WeakMap<object, string> = new WeakMap();

    /** Counter to get a new identifier */
    private declarationCounter: number = 0;

    /** Map of identifiers defined globally */
    private globalDeclarations: (
        | namedTypes.FunctionDeclaration
        | namedTypes.ClassDeclaration
        | namedTypes.VariableDeclaration
        | namedTypes.ExportNamedDeclaration
    )[] = [
        builders.exportNamedDeclaration(RequestErrorClass),
        builders.exportNamedDeclaration(ValidationErrorClass),
    ];

    /** Map of hashes already processed */
    private processedHashes: Set<string> = new Set();

    constructor(input: OpenAPISpec = {}) {
        this.input = input;
    }

    /**
     * Define a global identifier with a nickname.
     */
    public declareGlobally(
        declaration:
            | namedTypes.FunctionDeclaration
            | namedTypes.ClassDeclaration
            | namedTypes.VariableDeclaration,
    ) {
        this.globalDeclarations.push(declaration);
    }

    /**
     * Declare something globally basded on an input and return an identifier.
     */
    public declareForInput(
        input: any,
        gen: (
            id: namedTypes.Identifier,
        ) =>
            | namedTypes.FunctionDeclaration
            | namedTypes.ClassDeclaration
            | namedTypes.VariableDeclaration,
    ): namedTypes.Identifier {
        const hash = this.hashInput(input);
        const identifier = builders.identifier(hash);

        if (!this.processedHashes.has(hash)) {
            this.processedHashes.add(hash);
            const fn = gen(identifier);
            this.declareGlobally(fn);
        }

        return identifier;
    }

    /**
     * Define a function to validate an input.
     */
    public declareValidationFunction(
        input: object,
        gen: (args: {
            /** Identifier for the value argument being passed to the function */
            value: namedTypes.Identifier;
            /** Identifier for the path argument being passed to the function */
            path: namedTypes.Identifier;
            /** Identifier for the context argument being passed to the function */
            context: namedTypes.Identifier;
            /** Generate an error */
            error: (message: string) => namedTypes.NewExpression;
        }) => namedTypes.BlockStatement['body'],
    ) {
        const pathIdentifier = builders.identifier('path');
        const valueIdentifier = builders.identifier('value');
        const contextIdentifier = builders.identifier('context');

        const error = (message: string) => {
            return builders.newExpression(ValidationErrorIdentifier, [
                pathIdentifier,
                builders.literal(message),
            ]);
        };

        return this.declareForInput(input, (id) => {
            return builders.functionDeclaration(
                id,
                [pathIdentifier, valueIdentifier, contextIdentifier],
                builders.blockStatement(
                    gen({
                        value: valueIdentifier,
                        path: pathIdentifier,
                        context: contextIdentifier,
                        error,
                    }),
                ),
            );
        });
    }

    /**
     * Build the AST from the entire spec.
     */
    public indexAllComponents() {
        // Index all the schema components.
        const schemas = this.input.components?.schemas ?? {};
        Object.values(schemas).forEach((schema) => {
            compileValueSchema(this, schema);
        });
    }

    /**
     * Return the AST for the program.
     */
    public ast() {
        return builders.program([
            ...compileValidateRequest(this, this.input),
            ...this.globalDeclarations,
        ]);
    }

    /**
     * Generate the JS code for the AST.
     */
    public compile() {
        return escodegen.generate(this.ast(), {
            comment: true,
        });
    }

    /**
     * Hash an object and return an identifier name.
     */
    public hashInput(input: any): string {
        const isObject = typeof input === 'object' && input !== null;

        // Fast track for objects
        if (isObject && this.objectHashes.has(input)) {
            return this.objectHashes.get(input)!;
        }

        const hashValue = hash(input);
        if (this.hashes.has(hashValue)) {
            return this.hashes.get(hashValue)!;
        }

        const name = `obj${this.declarationCounter++}`;
        this.hashes.set(hashValue, name);
        if (isObject) {
            this.objectHashes.set(input, name);
        }

        return name;
    }

    /**
     * Resolve a reference to part of the spec.
     * We only support "#/" type of references.
     */
    public resolveRef(ref: OpenAPIRef) {
        const parts = ref.$ref.split('/').slice(1);

        let value: any = this.input;
        for (const part of parts) {
            value = value[part];
            if (value === undefined) {
                throw new Error(`Could not resolve reference ${ref.$ref}`);
            }
        }

        return value;
    }

    /**
     * Resolve a potential ref to a value.
     */
    public resolveMaybeRef<T>(value: T | OpenAPIRef): T {
        if (typeof value === 'object' && value !== null && '$ref' in value) {
            return this.resolveRef(value);
        }

        return value;
    }
}
