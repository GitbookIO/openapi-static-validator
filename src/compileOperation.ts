import { namedTypes, builders } from 'ast-types';

import { Compiler } from './compiler';
import { OpenAPIOperation } from './types';
import { compileValueSchema } from './compileValueSchema';
import { ValidationErrorIdentifier, buildRequestError } from './error';
import { OpenAPIParsedPath, getPathParamIndex, openapiPathToRegex } from './paths';

/**
 * Compile an operation into a function.
 * The value input is an object:
 * {
 *   path: string;
 *   method: string;
 *   body: any;
 *   query: any;
 *   headers: any;
 * }
 */
export function compileOperation(
    compiler: Compiler,
    path: OpenAPIParsedPath,
    operation: OpenAPIOperation,
) {
    return compiler.declareForInput(operation, (functionId) => {
        const requestIdentifier = builders.identifier('request');
        const pathMatchIdentifier = builders.identifier('pathMatch');
        const contextIdentifier = builders.identifier('context');

        const nodes: namedTypes.BlockStatement['body'] = [];

        // Set the operationId on the request
        if (operation.operationId) {
            nodes.push(
                builders.expressionStatement(
                    builders.assignmentExpression(
                        '=',
                        builders.memberExpression(
                            requestIdentifier,
                            builders.identifier('operationId'),
                        ),
                        builders.literal(operation.operationId),
                    ),
                ),
            );
        }

        // Extract and validate path params
        const pathParameters = (operation.parameters ?? []).filter(
            (parameter) => compiler.resolveMaybeRef(parameter).in === 'path',
        );
        nodes.push(
            builders.expressionStatement(
                builders.assignmentExpression(
                    '=',
                    builders.memberExpression(requestIdentifier, builders.identifier('params')),
                    builders.objectExpression(
                        pathParameters
                            .map((refParameter, index) => {
                                const parameter = compiler.resolveMaybeRef(refParameter);

                                const identifier = builders.identifier(`pathParam${index}`);
                                const schemaFn = compileValueSchema(compiler, parameter.schema);
                                const regexMatchIndex = getPathParamIndex(path, parameter.name);

                                nodes.push(
                                    builders.variableDeclaration('const', [
                                        builders.variableDeclarator(
                                            identifier,
                                            builders.callExpression(schemaFn, [
                                                builders.arrayExpression([
                                                    builders.literal('path'),
                                                    builders.literal(parameter.name),
                                                ]),
                                                builders.memberExpression(
                                                    pathMatchIdentifier,
                                                    builders.literal(regexMatchIndex),
                                                    true,
                                                ),
                                                contextIdentifier,
                                            ]),
                                        ),
                                    ]),
                                );

                                // Return an error if the parameter is invalid
                                nodes.push(
                                    builders.ifStatement(
                                        builders.binaryExpression(
                                            'instanceof',
                                            identifier,
                                            ValidationErrorIdentifier,
                                        ),
                                        builders.blockStatement([
                                            builders.returnStatement(identifier),
                                        ]),
                                    ),
                                );

                                return builders.property(
                                    'init',
                                    builders.identifier(parameter.name),
                                    identifier,
                                );
                            })
                            .flat(),
                    ),
                ),
            ),
        );

        // Validate query parameters
        const queryParameters = (operation.parameters ?? []).filter(
            (parameter) => compiler.resolveMaybeRef(parameter).in === 'query',
        );
        queryParameters.forEach((refParameter, index) => {
            const parameter = compiler.resolveMaybeRef(refParameter);
            const paramValueIdentifier = builders.identifier(`queryParam${index}`);
            const resultIdentifier = builders.identifier(`queryParamResult${index}`);
            const schemaFn = compileValueSchema(compiler, parameter.schema);

            // Assign the query parameter to a variable
            nodes.push(
                builders.variableDeclaration('const', [
                    builders.variableDeclarator(
                        paramValueIdentifier,
                        builders.memberExpression(
                            builders.memberExpression(
                                requestIdentifier,
                                builders.identifier('query'),
                            ),
                            builders.literal(parameter.name),
                            true,
                        ),
                    ),
                ]),
            );

            nodes.push(
                builders.ifStatement(
                    builders.binaryExpression(
                        '===',
                        paramValueIdentifier,
                        builders.identifier('undefined'),
                    ),
                    builders.blockStatement(
                        parameter.required
                            ? [
                                  builders.returnStatement(
                                      buildRequestError(
                                          400,
                                          `query parameter ${parameter.name} is required`,
                                      ),
                                  ),
                              ]
                            : [],
                    ),
                    builders.blockStatement([
                        // Validate the value
                        builders.variableDeclaration('const', [
                            builders.variableDeclarator(
                                resultIdentifier,
                                builders.callExpression(schemaFn, [
                                    builders.arrayExpression([
                                        builders.literal('query'),
                                        builders.literal(parameter.name),
                                    ]),
                                    paramValueIdentifier,
                                    contextIdentifier,
                                ]),
                            ),
                        ]),
                        // Fail if error
                        builders.ifStatement(
                            builders.binaryExpression(
                                'instanceof',
                                resultIdentifier,
                                ValidationErrorIdentifier,
                            ),
                            builders.blockStatement([builders.returnStatement(resultIdentifier)]),
                        ),
                        // Otherwise assign the value
                        builders.expressionStatement(
                            builders.assignmentExpression(
                                '=',
                                builders.memberExpression(
                                    builders.memberExpression(
                                        requestIdentifier,
                                        builders.identifier('query'),
                                    ),
                                    builders.literal(parameter.name),
                                    true,
                                ),
                                resultIdentifier,
                            ),
                        ),
                    ]),
                ),
            );
        });

        // Validate the body against the schema
        if (operation.requestBody) {
            const contentTypeSchema = operation.requestBody.content?.['application/json']?.schema;
            const bodyFn = contentTypeSchema
                ? compileValueSchema(compiler, contentTypeSchema)
                : null;
            const bodyResult = builders.identifier('body');

            nodes.push(
                builders.ifStatement(
                    builders.binaryExpression(
                        '===',
                        builders.memberExpression(requestIdentifier, builders.identifier('body')),
                        builders.identifier('undefined'),
                    ),
                    builders.blockStatement(
                        operation.requestBody.required
                            ? [builders.returnStatement(buildRequestError(400, 'body is required'))]
                            : [],
                    ),
                    builders.blockStatement(
                        bodyFn
                            ? [
                                  builders.variableDeclaration('const', [
                                      builders.variableDeclarator(
                                          bodyResult,
                                          builders.callExpression(bodyFn, [
                                              builders.arrayExpression([builders.literal('body')]),
                                              builders.memberExpression(
                                                  requestIdentifier,
                                                  builders.identifier('body'),
                                              ),
                                          ]),
                                      ),
                                  ]),
                                  builders.ifStatement(
                                      builders.binaryExpression(
                                          'instanceof',
                                          bodyResult,
                                          ValidationErrorIdentifier,
                                      ),
                                      builders.blockStatement([
                                          builders.returnStatement(bodyResult),
                                      ]),
                                      builders.blockStatement([
                                          builders.expressionStatement(
                                              builders.assignmentExpression(
                                                  '=',
                                                  builders.memberExpression(
                                                      requestIdentifier,
                                                      builders.identifier('body'),
                                                  ),
                                                  bodyResult,
                                              ),
                                          ),
                                      ]),
                                  ),
                              ]
                            : [],
                    ),
                ),
            );
        } else {
            nodes.push(
                builders.ifStatement(
                    builders.binaryExpression(
                        '!==',
                        builders.memberExpression(requestIdentifier, builders.identifier('body')),
                        builders.identifier('undefined'),
                    ),
                    builders.blockStatement([
                        builders.returnStatement(buildRequestError(400, 'body is not allowed')),
                    ]),
                ),
            );
        }

        nodes.push(builders.returnStatement(requestIdentifier));

        return builders.functionDeclaration(
            functionId,
            [requestIdentifier, pathMatchIdentifier, contextIdentifier],
            builders.blockStatement(nodes),
        );
    });
}
