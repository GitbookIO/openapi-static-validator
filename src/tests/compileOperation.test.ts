import { Compiler } from '../compiler';
import { compileOperation } from '../compileOperation';

describe('With body', () => {
    test('without', () => {
        const compiler = new Compiler();
        compileOperation(compiler, {});
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('required', () => {
        const compiler = new Compiler();
        compileOperation(compiler, {
            requestBody: {
                required: true,
                content: {
                    'application/json': {
                        schema: {
                            type: 'object',
                            properties: {
                                foo: {
                                    type: 'number',
                                },
                            },
                        },
                    },
                },
            },
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});
