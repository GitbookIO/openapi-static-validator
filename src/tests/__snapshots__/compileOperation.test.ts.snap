// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`With body without 1`] = `
"export function validateRequest(request) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value) {
    if (value.body !== undefined) {
        return new ValidationError(path, 'body is not allowed');
    }
    return value;
}"
`;

exports[`With body required 1`] = `
"export function validateRequest(request) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj2(path, value) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj1(path, value) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['foo'];
    if (value0 !== undefined) {
        const result0 = obj2([
            ...path,
            'foo'
        ], value0);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['foo'] = result0;
        keys.delete('foo');
    }
    if (keys.size > 0) {
        return new ValidationError(path, 'Unexpected properties');
    }
    return value;
}
function obj0(path, value) {
    if (value.body === undefined) {
        return new ValidationError(path, 'body is required');
    }
    const body = obj1([
        ...path,
        'body'
    ], value.body);
    if (body instanceof ValidationError) {
        return body;
    } else {
        value.body = body;
    }
    return value;
}"
`;
